service/tablet_allocator.cc

```diff
@@ -212,42 +288,46 @@ class load_balancer {
         // If target node is not balanced across shards, we will overload some shards.
         // If tablets are not balanced in size, throughput will suffer because some shards will be idle sooner than others.
         //
-        // FIXME: To handle the above, we should (1) rebalance the target node
-        // before migrating tablets from other nodes. If shards are balanced on the target node, the balancer
-        // will naturally distribute tablets to different shards. Also, (2) we should change this algorithm
-        // to be a generator for migrations and have a scheduler in the execution layer which pulls migrations
-        // from this algorithm, batches them and decides how many to execute.
-        //
-        // The scheduler decides in which order to execute the plan based on current activity in the system.
-        // We cannot just ask the planner for the next migration and stop when we hit overload on some shard,
-        // because that can lead to underutilization of the cluster. Just because the next migration is blocked
-        // by the target shard being busy doesn't mean we could not proceed with migrations for other shards
-        // which would be produced by the planner subsequently.
+        // FIXME: To handle the above, we should rebalance the target node before migrating tablets from other nodes.
+        // If shards are balanced on the target node, the balancer will naturally distribute tablets to different shards.
```

Comment:
If there is mild imbalance, won't it be corrected by the algorithm? It will choose less-loaded shards first.

We should reject a shard as a target if its load >= average load, as migrating a tablet to it will just cause it to get migrated back later.

Reply:
> If there is mild imbalance, won't it be corrected by the algorithm? It will choose less-loaded shards first.

Yes, it will, we fill least-loaded shards first. 

The problem is that imbalance will mean that we will not choose more loaded shards, so streaming capacity is reduced as those shards don't participate. If we equalized intra-node load first, which should be cheap, all shards could participate.

If the imbalance is mild and is corrected quickly, it's not an issue.

Reply:
Sure, but I see it as a second-order problem. Internal imbalance will be due to resharding, which is rare.

btw, we should aim to have good distribution of tables. Avoid table A on shard 0 and table B on shard 1, and instead have AB balanced on 01, in proportion to their sizes (and a similar policy for nodes).

Reply:
Yes, it is a second-order problem. It will get more serious when we base load on dynamic metrics like hotness.